<a href="https://colab.research.google.com/github/tiendatngcs/NLP/blob/master/4_1_Naive_Bayes_classifier_Vietnamese_news_topic_classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
import numpy as np
import pandas as pd
import pickle as pkl
from sklearn.naive_bayes import GaussianNB
import re
import string

In [3]:
with open("/content/drive/My Drive/dtm_6.pkl", "rb") as infile:
  (count_vectorizer, dtm, label) = pkl.load(infile)

In [4]:
df = pd.DataFrame(data=dtm.toarray(), columns=count_vectorizer.get_feature_names())
df["label"] = label

In [5]:
df

aaron  abbey  abc  abidal  aboluowang  ac  ...  ủi  ủng  ủy  ức  ứng  label
0         0      0    0       0           0   0  ...   0    0   0   0    0      0
1         0      0    0       0           0   0  ...   0    0   0   0    0      0
2         0      0    0       0           0   0  ...   0    0   0   0    0      0
3         0      0    0       0           0   0  ...   0    1   1   0   14      0
4         0      0    0       0           0   0  ...   0    0   0   0    0      0
...     ...    ...  ...     ...         ...  ..  ...  ..  ...  ..  ..  ...    ...
3637      0      0    0       0           0   0  ...   0    0   1   0   10      6
3638      0      0    0       0           0   0  ...   0    1   0   0    0      6
3639      0      0    0       0           0   0  ...   0    0   0   0    2      6
3640      0      0    0       0           0   0  ...   0    0   0   0    0      6
3641      0      0    0       0           0   0  ...   0    0   0   0    0      6

[3642 rows x 4828 columns]

In [6]:
topics = ['business', 'entertainment', 'sports', 'health', 'life', 'travel', 'digital']

In [33]:
X = dtm.toarray()
y = label

In [7]:
GNB = GaussianNB()

In [8]:
GNB.fit(X=dtm.toarray(), y=label)

GaussianNB(priors=None, var_smoothing=1e-09)

In [9]:
test_text = "Thẻ HDBank MasterCard với những ưu đãi nổi trội mang đến những lợi ích 'siêu hấp dẫn' không chỉ về phí mà còn nhận quà trong chương trình 'Khách hàng vay - tặng ngay máy lọc không khí' khi thực hiện giao dịch thanh toán mua hàng. ‘Khách hàng vay - tặng ngay máy lọc không khí’ cùng thẻ HDBank MasterCard. Từ nay đến ngày 31.10.2020, 180 khách hàng đầu tiên khi mở mới thẻ tín dụng quốc tế HDBank MasterCard và có chi tiêu phát sinh từ 20.000.000 đồng trong vòng 30 ngày kể từ ngày phát hành thẻ sẽ được tặng ngay máy lọc không khí. Thẻ tín dụng quốc tế HDBank MasterCard là dòng thẻ mà HDBank hướng đến trong việc đáp ứng đầy đủ các nhu cầu của khách hàng về du lịch, ăn uống, mua sắm và chi tiêu online. Với ưu đãi lên đến 14 triệu đồng/năm và miễn lãi lên đến 45 ngày sẽ mang tới cho chủ thẻ nhiều trải nghiệm hữu ích và thiết thực hơn. Khách hàng có thể thoải mái sử dụng thẻ tín dụng quốc tế HDBank MasterCard để giải quyết những vấn đề tài chính cá nhân mà vẫn có thể giảm bớt gánh nặng trong thanh toán. Bên cạnh đó, khi khách hàng sở hữu tấm thẻ này sẽ nhận được nhiều lợi ích từ chuỗi ưu đãi “khủng” khi kết hợp mua hàng và thanh toán trên các trang mua sắm trực tuyến, đối tác của HDBank như Tiki, Lazada, Shopee, Samsung... Với những đặc tính nổi trội, Thẻ tín dụng quốc tế HDBank MasterCard là “bí kíp” tiêu dùng hữu ích cho người trẻ trong cuộc sống hiện đại. Đặc biệt, vào thời điểm dịch Covid-19 bùng phát, chủ thẻ hoàn toàn có thể thực hiện giao dịch online mà không cần tập trung nơi đông người để mua sắm, thanh toán trực tiếp. Dùng thẻ để thanh toán chính là một phương thức thanh toán hiện đại và thẻ tín dụng quốc tế HDBank MasterCard là một lựa chọn phù hợp cho khách hàng bởi tính an toàn, tiện lợi và độ bảo mật cao với chuẩn chip EMV và công nghệ 3D Secure cho các giao dịch online. Đặc biệt quý khách hàng không còn phải lo lắng chuyển đổi ngoại tệ khi sử dụng thẻ để thanh toán khi đi công tác hoặc du lịch nước ngoài."

In [11]:
test_x = count_vectorizer.transform([test_text]).toarray()

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:385: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['bao', 'bay', 'biến', 'biệt', 'bây', 'bõm', 'bảo', 'bất', 'bẩy', 'bập', 'bắt', 'bội', 'chao', 'chi', 'chia', 'chu', 'chui', 'chuẩn', 'chà', 'chành', 'chí', 'chót', 'chùn', 'chăn', 'chũn', 'chưng', 'chạnh', 'chả', 'chầm', 'chầy', 'chập', 'chắn', 'chẳng', 'chết', 'chốc', 'chừ', 'chừng', 'coi', 'cu', 'cá', 'câu', 'cóc', 'công', 'cạnh', 'cảm', 'cầu', 'cật', 'cắt', 'cổ', 'cụ', 'cục', 'cực', 'da', 'dà', 'dĩ', 'dưng', 'dần', 'dầu', 'dịp', 'dở', 'dụng', 'gian', 'giá', 'giác', 'giời', 'ha', 'hiện', 'hoàn', 'hèn', 'hình', 'hô', 'hầu', 'hậu', 'hẳn', 'hồ', 'hỗ', 'hội', 'hợp', 'hự', 'khói', 'khô', 'khăn', 'khắc', 'khẳng', 'kia', 'kiện', 'kê', 'kì', 'kìa', 'kỳ', 'lai', 'le', 'liên', 'liệt', 'loạt', 'luận', 'luật', 'luốt', 'lình', 'lí', 'lô', 'lý', 'lập', 'lẽ', 'lị', 'lự', 'lực', 'mày', 'mòi', 'mù', 'mạng',

In [12]:
GNB.predict(test_x)

array([0])

# K-fold Validation

In [31]:
from sklearn import metrics, model_selection

In [48]:
kf = model_selection.KFold(n_splits=20, shuffle=True, random_state=42)

In [49]:
accuracies = []
for train_index, test_index in kf.split(X):
  X_train = X[train_index]
  X_test = X[test_index]
  y_train = np.array(y)[train_index]
  y_test = np.array(y)[test_index]
  GNB.fit(X_train, y_train)
  y_pred = GNB.predict(X_test)
  accuracies.append(metrics.accuracy_score(y_test, y_pred))
print(np.mean(accuracies))

0.7894013090734402


Guassian Naive Bayes model has the accuracy score of **78.94%**

In [59]:
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size= 0.2, random_state=42)
GNB.fit(X_train, y_train)
y_pred = GNB.predict(X_test)
print(metrics.classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.63      0.75       133
           1       0.76      0.78      0.77       101
           2       0.98      0.96      0.97       113
           3       0.79      0.78      0.79       102
           4       0.61      0.76      0.68       104
           5       0.58      0.60      0.59        86
           6       0.80      0.91      0.85        90

    accuracy                           0.78       729
   macro avg       0.78      0.78      0.77       729
weighted avg       0.79      0.78      0.78       729



In [43]:
text = "Tăng nguy cơ xuất huyết, chảy máu Theo TS-BS Nguyễn Văn Lâm, Giám đốc Trung tâm y học lâm sàng các bệnh nhiệt đới trẻ em, Bệnh viện Nhi T.Ư, có xu hướng tăng các trẻ mắc sốt xuất huyết nhập viện. Khởi đầu, trẻ mắc sốt xuất huyết có sốt cao và thường sốt cao trong 2 - 3 ngày đầu, có thể kèm theo ban xuất huyết trên da, đau bụng, nôn ói, kém ăn, tiểu ít... Các trường hợp nặng có thể xuất huyết tiêu hóa, rối loạn ý thức. TS-BS Nguyễn Văn Lâm cho biết: “Sốt xuất huyết gây sốt cao, có thể lên đến 39 - 40 độ C, cần hạ sốt kịp thời cho trẻ. Tuy nhiên, cần tránh dùng thuốc aspirin và ibuprofen”. Thuốc có tác dụng hạ sốt sâu, kéo dài nhưng có nhiều tác dụng phụ, ngăn sự tập kết tiểu cầu, chống đông máu khiến bệnh nhân tăng nguy cơ chảy máu. Aspirin còn có tác dụng phụ gây loét dạ dày tá tràng và có thể gây xuất huyết dạ dày, ói ra máu. “Bệnh nhân sốt xuất huyết nếu uống 2 thuốc trên, ở thể nhẹ có thể chỉ xuất huyết dưới da, thể nặng sẽ làm tăng nguy cơ chảy máu răng, chảy máu cam, thậm chí nôn ra máu, chảy máu ồ ạt”, BS Lâm khuyến cáo. BS Lâm hướng dẫn: Trong vụ dịch sốt xuất huyết, nên hạ sốt cho trẻ nhỏ bằng paracetamol, nhưng phải dùng đúng liều lượng (10 - 15 mg/kg cân nặng) và mỗi liều uống cách 4 - 6 tiếng. Song song với hạ sốt, cha mẹ cần để trẻ nghỉ ngơi ở nơi thoáng mát, ăn uống hợp vệ sinh, đủ dinh dưỡng, uống đủ nước. Nếu trường hợp hạ sốt không hiệu quả, cần đưa trẻ đến bệnh viện, tuyệt đối không tự ý tăng liều hoặc đổi thuốc hạ sốt với mong muốn tác dụng “mạnh hơn”, vì có thể gây ngộ độc do quá liều hoặc dùng thuốc không phù hợp. Không chủ quan với paracetamol Theo TS-BS Nguyễn Trung Nguyên, Giám đốc Trung tâm chống độc, Bệnh viện Bạch Mai (Hà Nội): “Paracetamol là thuốc không kê đơn, nhưng có thể sẽ là thuốc gây ngộ độc thường gặp nhất trong các tân dược vì người bị đau, sốt đều dùng. Paracetamol là thuốc hạ sốt an toàn cho người mắc sốt xuất huyết nhưng vẫn có thể gây ngộ độc nếu dùng không đúng”. BS Nguyên cho biết, người dân ngộ độc thuốc paracetamol thường do tự dùng thuốc sai dẫn đến quá liều, do quá liều lặp lại nhiều lần. Các trường hợp ngộ độc hay xảy ra khi sốt cao, sốt dai dẳng kéo dài, người bị đau kéo dài. “Người dùng paracetamol có xu hướng không quan tâm đến liều lượng, đặc biệt là tình huống “uống xong không đỡ lại uống thêm”. Nhưng vấn đề là tình trạng này lặp đi lặp lại nhiều lần, không để ý đến khoảng cách giữa các liều cho phép, không để ý đến tổng liều tối đa được phép, dẫn tới quá liều gây ngộ độc”, BS Nguyên lưu ý. Chuyên gia chống độc khuyến cáo: Sử dụng thuốc giảm đau, dù là thuốc không kê đơn, vẫn luôn cần tuân thủ đúng liều trong mỗi lần uống, khoảng cách giữa các lần uống. Các trường hợp có bệnh lý, có sử dụng thuốc điều trị bệnh khác, cần thông báo cho bác sĩ để được hướng dẫn sử dụng thuốc hạ sốt an toàn, hợp lý."

In [14]:
def clean_text(text: str):
  text = text.lower()  # lowercase
  text = re.sub('[%s]' % re.escape(string.punctuation), '', text)  # remove punctuations
  text = re.sub('\w*\d\w*', '', text)  # Remove digits
  text = re.sub('[‘’“”…]', '', text)
  text = re.sub('\n', ' ', text)
  text = re.sub('\s\s+', ' ', text)
  return text

In [19]:
clean_text = clean_text(text)

In [23]:
test_x = count_vectorizer.transform([clean_text]).toarray()

In [27]:
y_pred_on_train = GNB.predict(dtm.toarray())

In [30]:
metrics.accuracy_score(y_true=label, y_pred=y_pred_on_train)

0.9472817133443163